[DebugInfo][RemoveDIs] Use iterator-inserters in clang

jmorse opened the pull request:
As part of the LLVM effort to eliminate debug-info intrinsics, we're moving to a world where only iterators should be used to insert instructions. This isn't a problem in clang when instructions get generated before any debug-info is inserted, however we're planning on deprecating and removing the instruction-pointer insertion routines.

Scatter some calls to getIterator in a few places, remove a deref-then-addrof on another iterator, and add an overload for the createLoadInstBefore utility. Some callers passes a null insertion point, which we need to handle explicitly now.

jmorse committed:
[DebugInfo][RemoveDIs] Use iterator-inserters in clang

As part of the LLVM effort to eliminate debug-info intrinsics, we're moving
to a world where only iterators should be used to insert instructions. This
isn't a problem in clang when instructions get generated before any
debug-info is inserted, however we're planning on deprecating and removing
the instruction-pointer insertion routines.

Scatter some calls to getIterator in a few places, remove a
deref-then-addrof on another iterator, and add an overload for the
createLoadInstBefore utility. Some callers passes a null insertion point,
which we need to handle explicitly now.

OCHyams commented on clang/lib/CodeGen/CGCleanup.cpp:
```diff
@@ -295,18 +295,25 @@ void EHScopeStack::Cleanup::anchor() {}
                                   CodeGenFunction &CGF) {
-  auto store = new llvm::StoreInst(value, addr.emitRawPointer(CGF), beforeInst);
+  auto store = new llvm::StoreInst(value, addr.emitRawPointer(CGF),
+                                   beforeInst->getIterator());
```
Is it worth updating the `beforeInst` parameter to a iterator instead of `->getIterator()`-ing here? That would match the `createLoadInstBefore` overload you've added in this patch too.

OCHyams reviewed (approved):
LGTM with nit + pls appease the the formatter

jmorse committed:
Implement some review feedback (see reviews),

Also fix a stupid build failure where I'd assumed a &* pattern on a
variable was dereferencing-addrofing an iterator. It's actually a
ValueHandle, so we have to call getIterator directly.

jmorse committed:
clang-format

jmorse commented:
(I can't replicate the buildkite windows failure locally, will commit and see what actually goes wrong)

s-barannikov commented on clang/lib/CodeGen/CGCleanup.cpp:
```diff
@@ -293,20 +293,26 @@ void CodeGenFunction::initFullExprCleanupWithFlag(RawAddress ActiveFlag) {
                             beforeInst);
 }
 
+static llvm::LoadInst *createLoadInstBefore(Address addr, const Twine &name,
```
Insert before what?

s-barannikov commented on clang/lib/CodeGen/CGCleanup.cpp:
```diff
@@ -293,20 +293,26 @@ void CodeGenFunction::initFullExprCleanupWithFlag(RawAddress ActiveFlag) {
 
 static llvm::LoadInst *createLoadInstBefore(Address addr, const Twine &name,
-                                            llvm::Instruction *beforeInst,
+                                            llvm::BasicBlock::iterator beforeInst,
```
Please capitalize variable names and apply clang-format, this line exceeds 80 column limit.

s-barannikov reviewed (commented):
A couple of remarks in case this patch gets reverted.

jmorse commented:
Clang-formatted the declaration I touched in d12250ca7be, the other matters were already present in the code so I've shied away from addressing them.
